aptos-move/framework/aptos-framework/sources/object.move

```diff
@@ -0,0 +1,433 @@
+///   * Data resides in resources which cannot be moved and therefore are always accessible
+///   * In contrast, storable struct can be placed into an inaccessible space
+/// * Objects can be the source of events
+///   * When data is mutated in a struct, ideally that struct would emit an event.
+///   * The object model allows for events to be directly coupled with data.
+/// * Objects definition / structure can prevented after creation
+///   * At creation, the creator has the ability to configure the account as desired
+///   * Afterward, the creator can delegate signer ability to continue to change definition
+///     but it can also be discarded or access prevented to ensure that an object maintains
+///     consistent shape.
+/// * Secure and guaranteed to not collide with existing accounts.
+///   * Each address is generated in a collision-resistant fashion guaranteeing legitimate
+///     accounts cannot be created within the same address.
+///   * Two objects cannot be co-located due to an address being able to only store a single Object
+///
+/// TODO:
+/// * There is no means to borrow an object or a reference to an object. We are exploring how to
+///   make it so that a reference to a global object can be returned from a function.
+module aptos_framework::object {
+    use std::bcs;
+    use std::error;
+    use std::hash;
+    use std::signer;
+    use std::vector;
+
+    use aptos_framework::account;
+    use aptos_framework::event;
+    use aptos_framework::from_bcs;
+    use aptos_framework::guid;
+
+    /// An object already exists at this address
+    const EOBJECT_EXISTS: u64 = 0;
+    /// An object of that type already exists at this address
+    const EOBJECT_TYPE_EXISTS: u64 = 1;
+    /// An object does not exist at this address
+    const EOBJECT_DOES_NOT_EXIST: u64 = 2;
+    /// The object does not have ungated transfers enabled
+    const ENO_UNGATED_TRANSFERS: u64 = 3;
+    /// The caller does not have ownership permissions
+    const ENOT_OBJECT_OWNER: u64 = 4;
+    /// The object does not allow for deletion
+    const ECANNOT_DELETE: u64 = 5;
+
+    /// Scheme identifier used to generate an object's address. This serves as domain separation to
+    /// prevent existing authentication key and resource account derivation to produce an object
+    /// address.
+    const OBJECT_ADDRESS_SCHEME: u8 = 0xFE;
+
+    #[resource_group_member(group = aptos_framework::object::ObjectGroup)]
+    struct Object has key {
+        guid_creation_num: u64,
+        owner: address,
+        allow_ungated_transfer: bool,
+        transfer_events: event::EventHandle<TransferEvent>,
+    }
+
+    #[resource_group(scope = global)]
+    struct ObjectGroup { }
+
+    /// Type safe way of designate an object as at this address
+    struct ObjectId has copy, drop, store {
+        inner: address,
+    }
+
+    /// This is a one time ability given to the creator to configure the object as necessary
+    struct CreatorRef has drop {
```

Comment:
Why? It is the ref produced during creation... and is no longer there afterward

Reply:
my point is this - you have these refs:

1. CreatorRef
2. TransferRef
3. DeleteRef

etc. The 2nd and 3rd are named after what the holder of the Ref can do with it, but the first one is named after the outcome of creating that Ref. In my understanding the object is already "created" beforehand and a side effect of that is that the caller gets the CreatorRef that allows him to configure the object but not "create" it (since it already has been created). My point is that name is kinda inconsistent with the other ones. CreatorRef doesn't allow you to create an object.

Reply:
I would argue the object isn't created until the CreatorRef has been dropped. I can add more comments here. I do see your point, but there was clear intent on my part. Naming it anything other than that leads to a different expectation on its life time. Thoughts?

Reply:
yeah I hear you, `AdminRef` doesn't indicate the transient nature of the Ref, so I'd need to think of another name - thinking...
On that note there are a few examples where creator_ref looks already kinda confusing lol:
<img width="593" alt="image" src="https://user-images.githubusercontent.com/1221897/215029588-7102bbe7-d39b-4a1f-80e4-51a8217191eb.png">

Reply:
you're going to have to explain to me why... because to me this is exactly the principle I had in mind :sob:

Reply:
how about: `InitRef` | `InitCapability` | `InitCap` ? 
Suggestion 2 and 3 intentionally avoid the term `Ref` to disambiguate it from the other store-able Ref structs whose nature is a bit different.

Reply:
> you're going to have to explain to me why... because to me this is exactly the principle I had in mind 😭

the example above is confusing (imo) because you are using a variable `creator` which is a signer and refers to the address of the actual creator. Then on the other hand you have a variable creator_ref which contains the ability to initialize/write etc. to a specific object(-account) instead of the creator account. So the creator_ref has kind of very little to do with the actual creator account (different addresses etc.)

Reply:
Init is that different to you than Creator?